File: solutions/erlang/difference-of-squares/1/src/difference_of_squares.erl
-module(difference_of_squares).

-export([difference_of_squares/1, square_of_sum/1, sum_of_squares/1]).

-spec difference_of_squares(pos_integer()) -> pos_integer().
difference_of_squares(N) -> square_of_sum(N) - sum_of_squares(N).

-spec square_of_sum(pos_integer()) -> pos_integer().
square_of_sum(N) -> 
  Sum = N * (N + 1) div 2,
  Sum * Sum.

-spec sum_of_squares(pos_integer()) -> pos_integer().
sum_of_squares(N) -> 
  lists:foldl(fun(X, Acc) -> Acc + X*X end, 0, lists:seq(1, N)).
  

File: solutions/cpp/difference-of-squares/1/difference_of_squares.cpp
#include "difference_of_squares.h"

namespace difference_of_squares {
    int square_of_sum(int n) {
        int s = 0;
        for (int i = 1; i <= n; i++) {
            s += i;
        }
        return s * s;
    }
    int difference(int n) {
        return square_of_sum(n) - sum_of_squares(n);
    }
    int sum_of_squares(int n) {
        int s = 0;
        for (int i = 1; i <= n; i++) {
            s += i * i;
        }
        return s;
    }
}  // namespace difference_of_squares
